src/pubsub/peer-streams.js

```diff
@@ -43,35 +46,41 @@ class PeerStreams extends EventEmitter {
     this.id = id
     /**
      * Established protocol
+     *
      * @type {string}
      */
     this.protocol = protocol
     /**
      * The raw outbound stream, as retrieved from conn.newStream
+     *
      * @private
-     * @type {DuplexIterableStream}
+     * @type {null|DuplexIterableStream}
      */
     this._rawOutboundStream = null
     /**
      * The raw inbound stream, as retrieved from the callback from libp2p.handle
+     *
      * @private
-     * @type {DuplexIterableStream}
+     * @type {null|DuplexIterableStream}
      */
     this._rawInboundStream = null
```

Comment:
```suggestion
     * @type {DuplexIterableStream}
     */
    this._rawInboundStream
```

dont init to null or else you will need to protect everywhere this is used, just https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#classes

Reply:
This is a good practise that we can use. In this case it is more complex because multiple inbound streams will be attached over time to the peer.
If we have it in the constructor when we create the peer, we need to take into account only follow up attachments which makes this logic more complex. IMHO we should not add complexity to the code to favour internal types

Reply:
this is basically just trading null for undefined does it make a big difference ?

Reply:
Do you mean changing this to to equal undefined? It is a js file, I cannot just do the suggested change

Reply:
I think making those types non optional as in `Stream` over `Stream|null` (or `Stream|undefined`)  is a bad idea is it will prevent type checker from identifying real errors. E.g. some method that assumes `_rawInboundStream` might be called before it is initialized and type checker will not remind you to check that value is there.

I am assuming however that root of the problem [boils down to this example](https://www.typescriptlang.org/play?#code/MYGwhgzhAEAKCm8BOBlALk+YC2MDeAsAFDTQD2ArmgEaUB2AJupjgFzR0UgjQA+0AdSQBLNGGoh4zLNmKlgZOhAwVgaMkgAUASmiESpaGgAWwiADpKNekwwzoAXg5cQc6AF83Ac3hpoZoVFxSWgdPTdDTDQKJDpoACEyMkkwOk0TM0sqWgpGaRxtN08DAHcRNHhQhjAxVgBVYTo0AA4AQSQkMABPXX1DfwAzUIBCDIsA8uD4XojDEyQyEo54JYBRDo1NAHIAOTJoMqCJSoU6Ong1YUUtwoNSYtmxrOtc2xZscx80QIqtbXNDhVNNUxLd+gB6cGGAB60OgAHlqAArC5+MzQAAOZCgwgkXWgW043C2RWI7iAA) (simplified so it's easy to throw into ts playground):

```ts
class PeerStreams {
  outboundStream: null | WritableStream
  constructor() {
    this.outboundStream = null
  }
  get isWritable () {
     return Boolean(this.outboundStream)
  }
  write (data:Uint8Array) {
    if (!this.isWritable) {
      throw new Error('No writable connection')
    }

    this.outboundStream.getWriter().write(data)
    //   ^^ Object is possibly 'null'
  }
}
```

Where it is clear that `this.outboundStream` isn't null because of `!this.isWritable` check above. Unfortunately TS can't (yet) derive that because it does not know about relation here. It could however be told to [recognize that as follows](https://www.typescriptlang.org/play?#code/MYGwhgzhAEAKCm8BOBlALk+YC2MDeAsAFDTQD2ArmgEaUB2AJupjgFzR0UgjQA+0AdSQBLNGGoh4zLNmKlgZOhAwVgaMkgAUASmiESpaGgAWwiADpKNekwwzoAXg5cQc6AF83Ac3hpoZoVFxSWgdPTdDTDQKJDp-CECxCXhNEzNtN08DAHcRNHhQhjAxVgBVYTo0AA4AQSQkMABPXX1DfwAzUIBCALzglLSIbRaIwxMkMmyOeCmAUXqNTQByADkyaFyg5OgFOjp4NWFFJYyDUizRwcsqWgpGaRxzHzRE5B1zTfzNIrFT8+ILkRdsp4ol+o5QsoWNhWABZJrUeBg5LadhQ+xmQR9bYOAB8bgAQmQyJIwHRNOjHlZbvc7DhTsQKvkkO0wMACvDGojkSFWuQbjYHjCsVtJEK+M5uADiIzKshWeyRUleW5qYK6cKeVINdKiEA):

```ts
class PeerStreams {
  outboundStream: null | WritableStream
  constructor() {
    this.outboundStream = null
  }
  get isWritable () {
     return isWritable(this)
  }
  write (data:Uint8Array) {
    if (!isWritable(this)) {
      throw new Error('No writable connection')
    }

    this.outboundStream.getWriter().write(data)
  }
}

const isWritable = (stream:MaybeWritable): stream is Writable =>
  Boolean(stream.outboundStream)

interface MaybeWritable {
  outboundStream: WritableStream | null
}

interface Writable {
  outboundStream: WritableStream
}
```

P.S: Unfortunately you `: stream is Writable` types annotations don't seem to be (yet) supported on getters, but I do hope that helper function is not a large price to pay for the added benefit, especially since they could probably be reused elsewhere.